src/Controls/src/Core/Label/Label.cs

```diff
@@ -411,16 +407,75 @@ public override IList<GestureElement> GetChildElements(Point point)
 
-		void IPaddingElement.OnPaddingPropertyChanged(Thickness oldValue, Thickness newValue)
+		void IPaddingElement.OnPaddingPropertyChanged(Thickness oldValue, Thickness newValue) =>
+			InvalidateMeasureIfLabelSizeable();
+
+		Font ITextStyle.Font => this.ToFont();
+
+		/// <summary>
+		/// This method prevents unnecessary measure invalidations when the label is not
+		/// sizeable. If the label has a fixed width and height, then no matter what the
+		/// text is, the label will never change size.
+		/// </summary>
+		void InvalidateMeasureIfLabelSizeable()
 		{
+			if (!IsLabelSizeable(this))
+				return;
+
 			InvalidateMeasureInternal(InvalidationTrigger.MeasureChanged);
 		}
 
-		Font ITextStyle.Font => this.ToFont();
+		/// <summary>
+		/// Determines if the label can grow in any direction based on the constraints. If the
+		/// label cannot grow in any direction, then we usually don't need to do anything.
+		/// </summary>
+		internal static bool IsLabelSizeable(Label label)
+		{
+			// Determine in which direction the label can grow/shrink.
+			var constraint = label.Constraint;
+			var isVerticallySizeable = (constraint & LayoutConstraint.VerticallyFixed) == 0;
+			var isHorizontallySizeable = (constraint & LayoutConstraint.HorizontallyFixed) == 0;
+			var isSizeable = isVerticallySizeable || isHorizontallySizeable;
+
+			// If the label cannot grow in any direction, then we usually don't need to do anything.
+			if (!isSizeable)
+				return false;
+
+			// The label may grow/shrink based on the constraints, so we may need to invalidate.
+			return true;
+		}
+
+		/// <summary>
+		/// Determines if the text has changed in a way that would require a measure invalidation.
+		/// Unlike FormattedText changes, Text changes may not always require invalidation because
+		/// the text size and spacing is all uniform. Formatted text may have a case where even
+		/// though the label is a single line, the font size of a span may cause the label to grow
+		/// vertically.
+		/// </summary>
+		internal static bool TextChangedShouldInvalidateMeasure(Label label)
+		{
+			// If the label cannot grow in any direction, then we don't need to invalidate.
+			var isSizeable = IsLabelSizeable(label);
+			if (!isSizeable)
+				return false;
+
+			// Determine if the label can grow vertically (wrapping means it may grow vertically).
+			var constraint = label.Constraint;
+			var breakMode = label.LineBreakMode;
+			var isHorizontallySizeable = (constraint & LayoutConstraint.HorizontallyFixed) == 0;
+			var isMultiline = breakMode == LineBreakMode.CharacterWrap || breakMode == LineBreakMode.WordWrap;
+			var isSingleLine = !isMultiline;
+
+			// If the label cannot grow horizontally and is only single line,
+			// then we don't need to invalidate since the only direction it can grow in
+			// is vertically but it never will.
+			if (!isHorizontallySizeable && isSingleLine)
+				return false;
```

Comment:
This condition is the new condition. Besides the check to make sure that if the width AND height are fixed, this check makes sure to only avoid a layout if the WIDTH is constrained. Previously the height was the constraint that was used however it was not correct. And even if it was correct in some cases, we cannot assume vertical and horizontal are mutually exclusive.